Interop/Runtime/src/jvm/kotlin/kotlinx/cinterop/JvmNativeMem.kt

```diff
@@ -95,17 +95,88 @@ internal object nativeMemUtils {
+    // Alignments are such that overhead is approx 10%.
+    private const val SmallChunksSizeAlignment = 8
+    private const val MediumChunksSizeAlignment = 256
+    private const val BigChunksSizeAlignment = 4096
+    private const val MaxSmallSize = ChunkBucketSize * SmallChunksSizeAlignment
+    private const val MaxMediumSize = ChunkBucketSize * MediumChunksSizeAlignment
+    private const val MaxBigSize = ChunkBucketSize * BigChunksSizeAlignment
+    private const val ChunkHeaderSize = 2 * Int.SIZE_BYTES // chunk size + alignment hop size.
+
+    private val smallChunks = LongArray(ChunkBucketSize)
+    private val mediumChunks = LongArray(ChunkBucketSize)
+    private val bigChunks = LongArray(ChunkBucketSize)
+
+    // Chunk layout: [chunk size,...padding...,diff to start,aligned data start,.....data.....]
     fun alloc(size: Long, align: Int): NativePointed {
-        val address = unsafe.allocateMemory(
-                if (size == 0L) 1L else size // It is a hack: `sun.misc.Unsafe` can't allocate zero bytes
-        )
+        val totalChunkSize = ChunkHeaderSize + size + align
+        val ptr = ChunkHeaderSize + when {
+            totalChunkSize <= MaxSmallSize -> allocFromFreeList(totalChunkSize.toInt(), SmallChunksSizeAlignment, smallChunks)
+            totalChunkSize <= MaxMediumSize -> allocFromFreeList(totalChunkSize.toInt(), MediumChunksSizeAlignment, mediumChunks)
+            totalChunkSize <= MaxBigSize -> allocFromFreeList(totalChunkSize.toInt(), BigChunksSizeAlignment, bigChunks)
+            else -> unsafe.allocateMemory(totalChunkSize).also {
+                // The actual size is not used. Just put value bigger than the biggest threshold.
+                unsafe.putInt(it, Int.MAX_VALUE)
+            }
+        }
+        val alignedPtr = alignUp(ptr, align)
+        unsafe.putInt(alignedPtr - Int.SIZE_BYTES, (alignedPtr - ptr).toInt())
+        return interpretOpaquePointed(alignedPtr)
+    }
+
+    private fun allocFromFreeList(size: Int, align: Int, freeList: LongArray): NativePtr {
+        val paddedSize = alignUp(size, align)
+        val index = paddedSize / align - 1
+        val chunk = freeList[index]
+        val ptr = if (chunk == 0L)
+            allocRaw(paddedSize)
+        else {
+            val nextChunk = unsafe.getLong(chunk)
+            freeList[index] = nextChunk
+            chunk
+        }
+        unsafe.putInt(ptr, paddedSize)
+        return ptr
+    }
+
+    private fun freeToFreeList(paddedSize: Int, align: Int, freeList: LongArray, chunk: NativePtr) {
+        require(paddedSize > 0 && paddedSize % align == 0)
+        val index = paddedSize / align - 1
+        unsafe.putLong(chunk, freeList[index])
+        freeList[index] = chunk
+    }
 
-        if (address % align != 0L) TODO(align.toString())
-        return interpretOpaquePointed(address)
+    private const val RawChunkSize: Long = 4 * 1024 * 1024
+    private val rawChunks = mutableListOf<NativePtr>()
```

Comment:
IIUC, allocated raw chunks aren't freed.
If this is true, then memory in fact leaks here, which might be especially undesirable when running the compiler in the Gradle daemon process.

Reply:
Damn, you're right, native memory is not freed automatically after finishing compilation.